src/m365/pa/commands/app/app-permission-list.ts

```diff
@@ -0,0 +1,131 @@
+    return 'Lists all permissions of a Power Apps app';
+  }
+
+  public defaultProperties(): string[] | undefined {
+    return ['roleName', 'principalId', 'principalType'];
+  }
+
+  constructor() {
+    super();
+
+    this.#initTelemetry();
+    this.#initOptions();
+    this.#initValidators();
+    this.#initOptionSets();
+  }
+
+  #initTelemetry(): void {
+    this.telemetry.push((args: CommandArgs) => {
+      Object.assign(this.telemetryProperties, {
+        asAdmin: !!args.options.asAdmin,
+        environmentName: typeof args.options.environment !== 'undefined',
+        roleName: typeof args.options.roleName !== 'undefined'
+      });
+    });
+  }
+
+  #initOptions(): void {
+    this.options.unshift(
+      {
+        option: '--appName <appName>'
+      },
+      {
+        option: '--asAdmin'
+      },
+      {
+        option: '-e, --environmentName [environmentName]'
+      },
+      {
+        option: '--roleName [roleName]',
+        autocomplete: this.allowedRoleNames
+      }
+    );
+  }
+
+  #initValidators(): void {
+    this.validators.push(
+      async (args: CommandArgs) => {
+        if (!validation.isValidGuid(args.options.appName)) {
+          return `${args.options.appName} is not a valid GUID.`;
+        }
+
+        if (args.options.roleName && !this.allowedRoleNames.includes(args.options.roleName)) {
+          return `${args.options.roleName} is not a valid roleName. Allowed values are ${this.allowedRoleNames.join('|')}`;
+        }
+
+        return true;
+      }
+    );
+  }
+
+  #initOptionSets(): void {
+    this.optionSets.push(
+      {
+        options: ['environmentName'],
+        runsWhen: (args) => args.options.asAdmin
+      }
+    );
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    if (this.verbose) {
+      logger.logToStderr(`Retrieving permissions for app ${args.options.appName}${args.options.roleName !== undefined && ` with rolename ${args.options.roleName}`}`);
+    }
+
+    const url = `${this.resource}/providers/Microsoft.PowerApps${args.options.asAdmin ? '/scopes/admin' : ''}${args.options.environmentName ? '/environments/' + formatting.encodeQueryParameter(args.options.environmentName) : ''}/apps/${args.options.appName}/permissions?api-version=2022-11-01`;
+
+    try {
+      let permissions = await odata.getAllItems<{ principalType: string, principalId: string, roleName: string, properties: { roleName: string, principal: { id: string, type: string } } }>(url);
+
+      if (args.options.roleName) {
+        permissions = permissions.filter(permission => permission.properties.roleName === args.options.roleName);
+      }
+
+      if (permissions.length > 0) {
+        permissions.forEach(permission => {
+          permission.roleName = permission.properties.roleName;
+          permission.principalId = permission.properties.principal.id;
+          permission.principalType = permission.properties.principal.type;
+        });
```

Comment:
Let's do this only when it's really needed (when the output should be trimmed).

Reply:
Hi @milanholemans 

I was wondering, what are you meaning exactle `when the output should be trimmed`.

Currently, we have the method `Cli.shouldTrimOutput`, which will only trim the output when we use `--output text`, but that results in the following when using `--output md` or `--output csv`

![image](https://github.com/pnp/cli-microsoft365/assets/9193870/c3a1e699-340a-4500-bb1b-e8057007226c)

In this case, wouldn't it be better if we do `args.options.output !== json`, then the output would be like this:

![image](https://github.com/pnp/cli-microsoft365/assets/9193870/ffc88bc5-66ca-4ae3-b4c5-ce92c5517d92)

Reply:
`shouldTrimOutput` was made so we didn't have to check each command separately when we had to modify the result. And I kinda agree that the CSV and MD output for this command is ... not so ideal indeed.
@pnp/cli-for-microsoft-365-maintainers what is the way to go right now? Just use `Cli.shouldTrimOutput(...)` and leave the CSV and MD output scarce as shown above?
@martinlingstuyl I know you changed some stuff about the output. Maybe you could clarify it a bit?

Reply:
Hi @milanholemans, @MathijsVerbeeck, we changed two things about the output: 

1. All columns should be shown in md and csv output. (They were trimmed before)
2. Do not show nested objects in csv and md uitput, unless expressly requested using jmespath.

In this scenario, the question is what you would want to show; if the nested properties are really important to always show, and you don't want to change the api responses in json mode, you could of course flatten the object when `output !== 'json'`, like Mathijs suggests.

Reply:
So that means that the function `Cli.shouldTrimOutput(...)` has no use anymore?

Reply:
I'm not sure if you can say that without looking at where it's used. There may be valid use cases.

Reply:
But in this instance at least, you don't need it.

Reply:
> I'm not sure if you can say that without looking at where it's used. There may be valid use cases.

At this point I think it has outlived its usefulness, unless we add `csv` and `md` again.